docs/architecture/ErrorReporting.md

```diff
@@ -0,0 +1,122 @@
+# Handling and reporting errors
+
+    WARNING: This document describes future plans for error handling.
+    The UI library does not (yet) implement error handling as described here.
+
+This document describes how errors are handled within the ACS UI library and how they are reported to the surrounding application.
+
+## Error sources
+
+There are two categories of errors, and they are handled differently.
+
+1. ACS backend errors: Errors returned by the REST calls to ACS backend services.
+
+   1. These can be transient or permanent (special case: permission error).
+
+2. UI errors: Programming errors in the UI library / errors returned from other browser APIs
+
+## ACS backend errors
+
+Handling ACS backend errors is challenging when using either the composites or the pure components with the react bindings provided by the ACS UI library:
+
+* ACS backend API calls are asynchronous
+* ACS backend API calls are often initiated outside of the React component lifecycle methods (e.g on interaction handler callbacks or `useEffect` blocks).
+
+But in both cases, all ACS backend API calls are mediated by the (chat and calling) stateful clients. Thus, the stateful client tees errors to a special field in the state:
+
+```ts
+interface CallClientState {
+    // ... other fields ...
+    errors: ErrorState[];
+}
+
+interface ErrorState {
+    /**
+     * Locally generated unique ID for error.
+     * Allows for de-duplicating similar errors from repeated API calls.
+     */
+    id: string;
+    sdkClientMethod: 'CallClient.createCallAgent' | 'CallClient.getDeviceManager'; // ... and many more.
+    error: Error;
+}
+```
+
+Stateful client API method calls continue to return errors (e.g. as a failed promise). So they can be handled directly as well.
+
+The stateful clients stores the most recent errors, evicting older errors if necessary. The number of errors stored can be configured, or set to -1 to disable deletion of older errors. e.g., the composites disable error eviction in the stateful client and delete older errors themselves.
```

Comment:
Let's look at a scenario.
-> Contoso makes a call 
-> SDK throws Error 
-> Contoso is using `on('error', callback)` to handle the error
-> Contoso detects and handles the error fixing the cause of error
-> Contoso makes a call again, and it is successful this time.

Questions:
1. Do we remove the error from error state in this case?
2. If yes, how are you envisioning the mechanism to work?
3. If not, does that mean that Contoso can end up finding that error in the state and getting confused?

Reply:
Thanks a lot for bringing that up. You're absolutely right, we need to define how errors are cleared.

I think errors should be automatically cleared by the stateful client on success, as you mentioned. Additionally, stateful client should expose a way to clear errors (e.g. when the `ErrorBar` is dimissed).

Updated the doc, please take a look.